display_list/display_list_builder.cc

```diff
@@ -621,9 +644,23 @@ void DisplayListBuilder::clipPath(const SkPath& path,
       return;
     }
   }
-  clip_op == SkClipOp::kIntersect  //
-      ? Push<ClipIntersectPathOp>(0, 1, path, is_aa)
-      : Push<ClipDifferencePathOp>(0, 1, path, is_aa);
+  if (clip_op == SkClipOp::kIntersect) {
+    Push<ClipIntersectPathOp>(0, 1, path, is_aa);
+    if (!current_layer_->clip_bounds.intersect(path.getBounds())) {
+      current_layer_->clip_bounds.setEmpty();
+    }
+  } else {
+    Push<ClipDifferencePathOp>(0, 1, path, is_aa);
+  }
+}
+SkRect DisplayListBuilder::getLocalClipBounds() {
+  SkM44 inverse;
+  if (current_layer_->matrix.invert(&inverse)) {
+    SkRect devBounds;
+    current_layer_->clip_bounds.roundOut(&devBounds);
```

Comment:
Why roundOut?

Reply:
The clip_bounds are sub-pixel accurate, but the actual pixels touched will be rounded out to a whole pixel which means any rendering op that touches any of these pixels might leak through into the resulting rendering since we don't actually do geometric clipping, so they must be accounted for in the returned bounds for culling analysis and they must be accounted for before the inverse transform is applied in the next line.